### DISCUSSION TITLE
Creating new content?

### DISCUSSION THREAD

**Discussion by ghost** [DavidRieman/WheelMUD#72] [Q&A]
I've spent a bit of time looking at the database and nosql schemas. It seems like most interesting data at this time is stored in the relational db and at the moment only the player hash is stored in the document store. 

To author new content is it expected that this will be inserted directly into the db or will there before admin commands to create new rooms, exists, doors, etc... from within the MUD interface itself? I see admin commands such as CreatePortal,
CreatePotion, and CreateWeapon which seem to infer the latter.

Thanks and if there's any world creation documentation I've missed sorry for wasting your time! :)

**Comment by DavidRieman** [comment]
Yeah pretty much the whole player object is stored in the document store but other data has not yet been migrated there.  The intent is to move all the "world" stuff into the document store as Behavior-decorated Things, leaving just a few things (like fast player lookup) to the relational table.  But things will get interesting as we look to things like mobs - wanting to store "spawner" behaviors on rooms and/or areas instead of storing mob instances directly.

Eventually it may turn out document DB lookups (e.g. RavenDB's auto-indexing) may be _fast enough_ to forego the relational DB entirely.  We'll see.

We've written a good deal on intended On-Line Creation (OLC) design thoughts, which has been high on my priority list to find and migrate here from the old forum backups for some time.  I've only recently whittled down thousands of old posts (and duplicates) from old forums backups to just ~1000 to go through by hand to recover this stuff.  The short of it though is that we do want to build OLC features but haven't started yet.  The recent migration to document DB though does feel like it warrants bringing that into play _sooner_ than we originally planned on the roadmap though, since building the data by hand directly into the DB isn't really viable.

The functions like CreatePortal/CreatePotion/CreateWeapon are all old commands that have been disabled either with or before the move to document DB.  In OLC, we'd want something more like an admin would build, in-game, a "template" that is a portal or potion or weapon, rather than separate hard-coded commands.

I'll try to resurrect our docs and discussions on OLC and point you to it when ready, to resume discussing where we want to take this.

**Reply by DavidRieman** [reply]
Success!  Initial OLC doc added to Design section at https://github.com/DavidRieman/WheelMUD/blob/master/Documentation/Design/README.md - though I'm missing one small section and notes, it's late so I figured to push it up anyway.  Check it out and let me know what you think.

**Comment by ghost** [comment]
The OLC user story looks good and makes the rational for choosing a nosql document store more apparent. Is the idea to allow content to be authored in bulk outside the CLI as importable JSON documents? Perhaps these could be validated against json schema which would form the basis for the Thing templates you mentioned?

**Reply by DavidRieman** [reply]
Yes, being able to export and import bulk objects (like Areas) is something we'll want to support some day for content sharing.  I'll jot a note for that on the 1.1+ road map for now, as that'll be sweet after the framework starts seeing some adoption.

Before we moved to document DB, we also had a prototype offline world creation UI.  I think that'd tie well into the import/export stuff and will note that at 1.1+ as well.  IIRC what we had tied directly to the SQLite DB to insert/edit rows for the stuff you were creating.  It was pretty tied to the DB.  What I envision for a future version of this would be more like OLC:  The code could use MEF and/or reflection to find all the Behaviors and such and their user-editable properties, but provides a nice clickable UI for selecting through things.  (Too far out to worry about the specifics yet, IMO.)

A big part of why Document DB and OLC being so dynamic is we want WheelMUD to be a highly composable set of building blocks to plug together new interesting world interaction with relatively little fuss, and even sharing those building blocks (often Behaviors) around the community.  Like imagine someone builds a DrivableBehavior, and you compose that with an EnterableExitableBehavior to make a Thing that is now essentially a Car.  Want the doors to be lockable?  Add the LocksUnlocksBehavior and ensure it's events look at Enter requests to cancel them if locked.  Etc.

For the templates I mentioned - I foresee the most typical Things that a given MUD wants to create being dependent upon their particular needs.  Like a typical fantasy MUD may have "weapons" and "armor" and "potions" and "wands" and so on - each of those might in turn be comprised of a combination of several behaviors that make it something you can pick up, wield, wear, drink, break, cast with, etc.  Instead of a CreateWeapon game command, the world designers would first create a "weapon" template that is maybe based off the "object" template but adds WeaponBehavior or WieldableBehavior.  (Our reference implementation would start with such template in the DB but would allow you to change it or make more, etc.)  Common templates in a space-faring genre might be "starship" and "planet" and "space station", etc.

So in the old OLC doc I resurrected there, I think now I would revise a few things:  The first "new object" is probably more suited as "new thing" which would be the absolute base "Make a blank Thing" that you couldn't modify or delete as a template.  Then an "object" template would be a Thing but it starts with a PickUpDropBehavior (maybe "ObtainableBehavior"?  Names are hard and we hadn't settled on strong conventions for this stuff yet).  Then a weapon template would be a Thing that starts with PickUpDropBehavior and WeaponBehavior... etc.

Ideally, thinking on it today, I think the "new object from object/123" example I had in the doc would more aptly be a command like "clone object/123" - and a "template" will just be like any other thing you can clone, but one an admin marked with a user-friendly name like "weapon" for wide re-use instead of "object/123".
